WordPress/Classes/Login/WordPressDotComAuthenticator.swift

```diff
@@ -9,46 +9,176 @@ import Alamofire
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "success"])
+            return account
+        } catch {
+            present(error, from: viewController)
+            WPAnalytics.track(.wpcomWebSignIn, properties: ["stage": "error", "error": "\(error)"])
+            return nil
+        }
+    }
+
+    /// Sign in WP.com account.
+    ///
+    /// - SeeAlso: `signIn`
+    ///
+    /// - Parameters:
+    ///   - email: When provided, the signed-in account must be the account with the given email address.
+    @MainActor
+    func attemptSignIn(from viewController: UIViewController, context: SignInContext) async throws(SignInError) -> TaggedManagedObjectID<WPAccount> {
+        let defaultAccount = try? WPAccount.lookupDefaultWordPressComAccount(in: coreDataStack.mainContext)
+        let hasAlreadySignedIn = defaultAccount != nil
+
         let token: String
         do {
-            token = try await authenticate(from: viewController)
+            token = try await authenticate(from: viewController, prefersEphemeralWebBrowserSession: hasAlreadySignedIn)
         } catch {
-            if let error = error as? WordPressDotComAuthenticator.Error {
-                presentSignInError(error, from: viewController)
-            } else {
-                wpAssertionFailure("WP.com web login failed", userInfo: ["error": "\(error)"])
-            }
-            return
+            throw .authentication(error)
         }
 
-        let delegate = WordPressAuthenticator.shared.delegate!
-        let credentials = AuthenticatorCredentials(wpcom: WordPressComCredentials(authToken: token, isJetpackLogin: false, multifactor: false))
         SVProgressHUD.show()
-        delegate.sync(credentials: credentials) {
+        defer {
             SVProgressHUD.dismiss()
+        }
+
+        // Fetch WP.com account details
+        let user: RemoteUser
+        do {
+            let service = AccountServiceRemoteREST(wordPressComRestApi: .defaultApi(oAuthToken: token, userAgent: WPUserAgent.defaultUserAgent()))
+            user = try await withCheckedThrowingContinuation { continuation in
+                service.getAccountDetails(success: { continuation.resume(returning: $0!) }, failure: { continuation.resume(throwing: $0!) })
+            }
+        } catch {
+            throw .fetchUser(error)
+        }
+
+        // Make sure the signed-in account matches the given `accountEmail` argument.
+        if let email = context.accountEmail(in: coreDataStack.mainContext), user.email != email {
+            throw .mismatchedEmail(expectedEmail: email)
+        }
+
+        if let defaultAccountEmail = defaultAccount?.email, defaultAccountEmail != user.email {
+            throw .alreadySignedIn(signedInAccountEmail: defaultAccountEmail)
+        }
+
+        // Save the signed-in account details (and sites) into Core Data.
+        let accountID: TaggedManagedObjectID<WPAccount>
+        do {
+            let isJetpackLogin: Bool
```

Comment:
The `isJetpackLogin` scenario is unclear. It looks like it's for logging in with an account different from the current one, which is not fully supported.

```swift
if isJetpackLogin && !account.isDefaultWordPressComAccount {
            let blogService = BlogService(coreDataStack: coreDataStack)
            blogService.associateSyncedBlogs(toJetpackAccount: account, success: onSuccessInternal, failure: onFailureInternal)
        }
```

Do you think we should keep it or remove the complexity?

<hr/>

Now seems like a good time to specify the requirements for self-hosted site, cc @jkmassel. I think a lot of the technical complexity is the result of the UX complexity. I'd start with the following questions:

- Why does the app allow you to add a self-hosted site to the list of your wp.com sites for your current account? What does this even operation even mean? Does it actually "add" the site to your wp.com account?
- How would've it looked like if the app supposed multiple wp.com accounts?
- What is the best UX for adding a self-hosted site to your wp.com account (adding Jetpack)?

<img width="320" alt="Screenshot 2024-11-05 at 8 37 11 AM" src="https://github.com/user-attachments/assets/d2987b27-ea70-44fb-81f5-dd627958a7ea">

I think there is an opportunity to eliminate the scenario where you are logged in to your wp.com account and, at the same time, have a self-hosted site you are trying to connect Jetpack to.

Reply:
> The isJetpackLogin scenario is unclear. It looks like it's for logging in with an account different from the current one, which is not fully supported.

It's related to multiple WP.com accounts support, but not entirely. The code you quoted is the part that's important to this PR. It controls how the sites are synced into the database. I don't think that part needs to be coupled with "jetpack login". It's worth a refactor for later.